src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+
+                int initializerCaptureId = _availableCaptureId++;
+                AddStatement(new FlowCapture(initializerCaptureId, objectCreation.Syntax, objectCreation));
 
-            // PROTOTYPE(dataflow): For a non-null initializer, we'll need to assign the created object to a flow reference and rewrite all initializers to assign to/call functions on it, then return return the flow reference
+                objectCreation = new FlowCaptureReference(initializerCaptureId, objectCreation.Syntax, objectCreation.Type, objectCreation.ConstantValue);
+                HandleObjectOrCollectionInitializer(operation.Initializer, objectCreation);
+            }
 
             return objectCreation;
         }
 
+        public override IOperation VisitMemberInitializer(IMemberInitializerOperation operation, int? captureIdForResult)
+        {
+            Debug.Assert(_currentImplicitReceiver != null);
+
+            // We explicitly do not push the initialized member onto the stack here. We visit the initialized member to get the implicit receiver that will be substitued in when an
+            // IInstanceReferenceOperation with InstanceReferenceKind.ImplicitReceiver is encounted. If that receiver needs to be pushed onto the stack, its parent will handle it.
+            // In member initializers, the code generated will evalute InitializedMember multiple times. For example, if you have the following:
+            //
+            // class C1
+            // {
+            //   public C2 C2 { get; set; } = new C2();
+            //   public void M()
+            //   {
+            //     var x = new C1 { C2 = { P1 = 1, P2 = 2 } };
+            //   }
+            // }
+            // class C2
+            // {
+            //   public int P1 { get; set; }
+            //   public int P2 { get; set; }
+            // }
+            //
+            // We generate the following code for C1.M(). Note the multiple calls to C1::get_C2().
+            //   IL_0000: nop
+            //   IL_0001: newobj instance void C1::.ctor()
+            //   IL_0006: dup
+            //   IL_0007: callvirt instance class C2 C1::get_C2()
+            //   IL_000c: ldc.i4.1
+            //   IL_000d: callvirt instance void C2::set_P1(int32)
+            //   IL_0012: nop
+            //   IL_0013: dup
+            //   IL_0014: callvirt instance class C2 C1::get_C2()
+            //   IL_0019: ldc.i4.2
+            //   IL_001a: callvirt instance void C2::set_P2(int32)
+            //   IL_001f: nop
+            //   IL_0020: stloc.0
+            //   IL_0021: ret
+            //
+            // We therefore visit the InitializedMember to get the implicit receiver for the contained initializer, and that implicit receiver will be cloned everywhere it encounters
+            // an IInstanceReferenceOperation with ReferenceKind InstanceReferenceKind.ImplicitReceiver
+
+            IOperation implicitReceiver = Visit(operation.InitializedMember);
```

Comment:
>implicitReceiver [](start = 23, length = 16)

The name is feels misleading, it looks like this is not a receiver, but an instance to initialize. Consider renaming.

Reply:
After looking at what HandleObjectOrCollectionInitializer is doing, the name doesn't sound that surprising any more, but it is still feels like this terminology is specific to internal workings of HandleObjectOrCollectionInitializer 

---
In reply to: [184474823](https://github.com/dotnet/roslyn/pull/26389#discussion_r184474823) [](ancestors = 184474823)